### PR TITLE
IPSec key rotation without agent restart

### PR DESCRIPTION
**Pull request by jibi**

### CODE REVIEW

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (christarazi) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:645]: Would be good to wire up daemon's `context` for this loop to exit on shutdown

File: pkg/nodediscovery/nodediscovery.go
Comment (kaworu) [inline pkg/nodediscovery/nodediscovery.go]: nit
```suggestion
		case <-time.After(defaults.NodeInitTimeout):
```

**Review by kaworu — COMMENTED** [review]
@jibi FWIW we have some wrapper and logic around fsnotify to watch over TLS certificates [here](https://github.com/cilium/cilium/tree/master/pkg/crypto/certloader). The reason I'm mentioning this is that watching over k8s secrets has proven to be tricky:

1. Because of the symlinks structure and replacement hapenning when the "underlying" secret change
2. Because we get "a lot of events" (i.e. more than one) when a secret change
3. Because certloader handle `optional` volumes (not the case for ipsec keys if I understand correctly)

For 1. a [`fswatcher.Watcher`](https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/crypto/certloader/fswatcher/fswatcher.go) could be re-used here. For 2. we could take inspiration of certloader's [event coalescing mechanism](https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/crypto/certloader/watcher.go#L132-L133).

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (gandro) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: Using fsnotify to watch for Kubernetes secret changes might not always work, due to it using symlinks (see #13621 for how to reproduce). Have you checked if you could use the `fswatcher` package instead?

https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/crypto/certloader/fswatcher/fswatcher.go#L25-L50

Edit: @kaworu beat me to it.

**Comment by jibi** [comment]
Thanks for the feedback @kaworu @gandro :pray:
I switched to `fswatcher` as indeed with my approach, when using a k8s secret, I was not seeing any event for the keyfile

File: pkg/nodediscovery/nodediscovery.go
Comment (pchaigno) [inline pkg/nodediscovery/nodediscovery.go:304]: I don't see any reference to this function in the commit description? Wrong commit maybe?

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: Nit:
```suggestion
			log.WithError(err).WithField(logfields.Path, keyfilePath).
 				Warning("error encountered while watching file with fsnotify")
```

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: Should we make this an Info message? I'm not sure if we have other log messages for when the secret changes.

**Review by pchaigno — APPROVED** [review]
Only minor comments. Nothing blocking.

Thanks for the nice commit history! :pray:

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (jibi) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: We actually have already this information logged in the `fswatcher` package:
```
level=debug msg="Received fsnotify event" file-path=/etc/ipsec/keys operation=CHMOD subsys=fswatcher
```

I think I'll get rid of this and add an `Info` logline to `LoadIPSecKeysFile()` so that we log events only when the keys get actually rotated

File: pkg/nodediscovery/nodediscovery.go
Comment (jibi) [reply pkg/nodediscovery/nodediscovery.go:304]: Forgot to update the commit message

**Review by kaworu — CHANGES_REQUESTED** [review]
Thanks @jibi, overall LGTM. Now an assumption made before this patch about `ipSecKeysGobal` being safe to read without locking don't hold anymore: https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/datapath/linux/ipsec/ipsec_linux.go#L45-L47

So I believe we have to:

1. add a `sync.RWMutex` protecting [`ipSecKeysGlobal`](https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/datapath/linux/ipsec/ipsec_linux.go#L45-L47)
2. update the comment
3. read-lock on [`getIPSecKeys`](https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/datapath/linux/ipsec/ipsec_linux.go#L49)
4. write-lock on [`loadIPSecKeys`](https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/pkg/datapath/linux/ipsec/ipsec_linux.go#L481)

I'm _guessing_ writing without locking in `pkg/datapath/linux/ipsec/ipsec_linux_test.go` is still ok though, but to be checked.

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (kaworu) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]:
```suggestion
		case <-ctx.Done():
			watcher.Stop()
			return
```

**Review by kaworu — APPROVED** [review]

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (kaworu) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]:
```suggestion
	defer ipSecKeysGlobalLock.RUnlock()
```

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (kaworu) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]:
```suggestion
	defer ipSecKeysGlobalLock.Unlock()
```

**Review by kaworu — CHANGES_REQUESTED** [review]

File: pkg/datapath/linux/ipsec/ipsec_linux.go
Comment (jibi) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: :man_facepalming:

**Review by kaworu — APPROVED** [review]

**Comment by jibi** [comment]
/test-1.23-net-next

**Review by gandro — APPROVED** [review]

**Review by ldelossa — APPROVED** [review]
LGTM!

**Review by christarazi — APPROVED** [review]

**Comment by jibi** [comment]
CI is green (just rebased yesterday on master) but can't approve it as tophat as I'm also the author, so marking it as ready

**Comment by joestringer** [comment]
@jibi I'm excluding this PR from the current round of releases since the backports don't seem to be ready at this time. When you get a chance, please make sure to update the backport PR's descriptions to match the common description that we use for backports, so that we get cross-link references from the backport to this PR, and so that we can correctly generate the release notes when this patch _does_ get included in a release.

Also, I dropped the 1.9 pending label since it looked like you closed it.

**Comment by jibi** [comment]
> @jibi I'm excluding this PR from the current round of releases since the backports don't seem to be ready at this time.

@joestringer Perfect, I added the `backport-pending` labels on purpose as I'm taking care of the backports but they are currently blocked on https://github.com/cilium/cilium/pull/19932. Once that gets merged I'll update https://github.com/cilium/cilium/pull/20127 with a proper PR description and prepare the 1.11 one

> Also, I dropped the 1.9 pending label since it looked like you closed it.

Thanks, I forgot about that yesterday when I closed it

**Pull request merged**
